README.md
<a href="http://imgur.com/KyBBOwk"><img src="http://i.imgur.com/KyBBOwk.png" title="source: Justin C. Bagley" width=60% height=60% align="center" /></a>

[![Codacy Badge](https://app.codacy.com/project/badge/Grade/b34c50c07d1e4bd999f8861955822134)](https://www.codacy.com/gh/justincbagley/piranha/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=justincbagley/piranha&amp;utm_campaign=Badge_Grade) [![License](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)

Shell script pipeline for inferring ML gene trees for many loci (e.g. RAD loci, UCEs)

## NOTICE

>**MAGNET is currently only being actively maintained and developed within [PIrANHA](https://github.com/justincbagley/piranha), where it is at [v1.2.0](https://github.com/justincbagley/piranha/tree/master/bin/MAGNET-1.1.0).** 
>**The present repository will be periodically updated to include a standalone version of the MAGNET software. This helps users interested _only_ in MAGNET; however, it is strongly recommended that you install the full PIrANHA distribution and request MAGNET updates or bug fixes through the [PIrANHA repo](https://github.com/justincbagley/piranha) (e.g. [contact the author](https://github.com/justincbagley/piranha/wiki/6.-Contact), or raise an issue [here](https://github.com/justincbagley/piranha/issues)).**

## LICENSE

All code within the PIrANHA repository, including MAGNET v1.2.0 pipeline code, is available 'AS IS' under a 3-clause BSD license. See the [LICENSE](https://github.com/justincbagley/piranha/blob/master/LICENSE) file for more information.

## CITATION

Should you cite **MAGNET**? See https://github.com/mr-c/shouldacite/blob/master/should-I-cite-this-software.md.

If you use scripts from this repository as part of your published research, then please cite MAGNET as well as the PIrANHA package that it comes distributed in, as follows (also see DOI information below):

-   Bagley, J.C. 2020. PIrANHA v0.4-alpha4. GitHub repository, Available at: http://github.com/justincbagley/piranha/.
-   Bagley, J.C. 2020. MAGNET v1.2.0. GitHub package, Available at: http://github.com/justincbagley/MAGNET. 

Alternatively, please provide the following links to these software program in your manuscript:

-   http://github.com/justincbagley/piranha
-   http://github.com/justincbagley/MAGNET
  
Example citation using the MAGNET URL:

-   "We estimated a gene tree for each RAD locus in RAxML v8 (Stamatakis 2014) using 
	the MAGNET v1.2.0 pipeline (http://github.com/justincbagley/MAGNET). Each RAxML run
	specified the GTRGAMMA model and coestimated the maximum-likelihood phylogeny and
	bootstrap proportions from 500 bootstrap pseudoreplicates."

Example citation using the MAGNET and PIrANHA URLs:

-   "We estimated a gene tree for each RAD locus in RAxML v8 (Stamatakis 2014) using 
	the MAGNET v1.2.0 pipeline (http://github.com/justincbagley/MAGNET), as distributed in
	PIrANHA v0.4a4 (http://github.com/justincbagley/piranha). Each RAxML run specified
	the GTRGAMMA model and coestimated the maximum-likelihood phylogeny and bootstrap
	proportions from 500 bootstrap pseudoreplicates."

## DOI

The DOI for MAGNET, via Zenodo, is as follows:  [![DOI](https://zenodo.org/badge/66839898.svg)](https://zenodo.org/badge/latestdoi/66839898). Here is an example of citing MAGNET using the DOI: 
  
-   Bagley, J.C. 2020. MAGNET v1.2.0. GitHub package, Available at: https://doi.org/10.5281/zenodo.596774.

## INTRODUCTION

The estimation of species-level phylogenies, or 'species trees' is a fundamental goal in evolutionary biology. However, while 'gene trees' estimated from different loci provide insight into the varying evolutionary histories of different parts of the genome, gene trees are random realizations of a stochastic evolutionary process. Thus, gene trees often exhibit conflicting topologies, being incongruent with each other and incongruent with the underlying species tree due to a variety of genetic and biological processes (e.g. gene flow, incomplete lineage sorting, introgression, selection). 

With the advent of recent advances in DNA sequencing technologies, biologists now commonly sequence data from multiple loci, and even hundreds to thousands of loci can quickly be sequenced using massively parallel sequencing on NGS sequencing platforms. Full-likelihood or Bayesian algorithms for inferring species trees and population-level parameters from multiple loci, such as \*BEAST and SNAPP, are computationally burdensome and may be difficult to apply to large amounts of data or distantly related taxa (or other cases that complicate obtaining MCMC convergence). By contrast, a number of recently developed and widely used 'summary-statistics' approaches rely on sets of gene trees to infer a species tree for a set of taxa (reviewed by Chifman and Kubatko, 2014; Mirarab and Warnow, 2015). These methods are specifically designed to estimate gene trees or use gene trees input by the user, which are treated as observed data points analyzed in a distance-based or coalescent algorithm. Moreover, summary-statistics approaches to species tree inference tend to be accurate and typically much faster than full-data approaches (e.g. Mirarab et al., 2014;  Chifman and Kubatko, 2014). Examples of species tree software in this category include programs such as BUCKy (Larget et al., 2010), STEM (Liu et al., 2010), spedeSTEM, NJst (Liu and Yu, 2011), ASTRAL and ASTRAL-II (Mirarab and Warnow, 2015), and ASTRID (Vachaspati and Warnow, 2015). Phylogenetic network models implemented in recent software like SplitsTree and SNaQ also improve network and inference by analyzing sets of gene trees. 

Despite the importance of gene trees in species tree and network inference, few resources have been specifically designed to aid rapid estimation of gene trees for different loci. MAGNET (MAny GeNE Trees) is a shell script pipeline within the [PIrANHA](https://github.com/justincbagley/piranha) software package (Bagley 2020) that helps fill this gap by automating inferring a maximum-likelihood (ML) gene tree for each locus in a multilocus dataset. Here, the term "locus" is used loosely to refer to a DNA alignment of homologous nucleotide characters including both variable and invariant DNA sites. The MAGNET package was originally coded up to aid analyses of RAD loci generated by massively parallel sequencing of ddRAD-seq genomic libraries (Peterson et al., 2012). However, MAGNET can be used to estimate gene trees from any multilocus dataset in the appropriate format, and three starting file types are supported: single NEXUS, single G-PhoCS, or multiple PHYLIP files.

## HARDWARE AND SETUP

:computer: MAGNET focuses on allowing users to automate the workflow necessary for quickly estimating gene trees for many loci on their local machine. 

:thumbsup: No special hardware or setup is necessary, unless the user is interested in estimating gene trees on a remote supercomputing cluster (see below).

### Installation

Downloading, installing, and upgrading MAGNET (on macOS or Linux) is easy thanks to [Homebrew](https://brew.sh) distribution.

#### A. Homebrew install

[See here.](https://github.com/justincbagley/homebrew-tap)

Two Homebrew Install Options:

**Recommended:**

1. Development (cutting edge) Homebrew install:

```bash
# Install:
brew tap justincbagley/homebrew-tap ;
brew update ;
brew install --HEAD magnet ;
piranha -i ;
```

2. Regular Homebrew install (stable, but may not be up to date):

```bash
# Install:
brew tap justincbagley/homebrew-tap ;
brew update ;
brew install magnet ;
```
After running one of the installs *above*, it is recommended that users do `source ~/.bash_profile` next on **macOS**, or `source ~/.bashrc` next on **Linux**. 

#### B. Regular installation 'by-hand':

You can also install MAGNET manually by

1.  conducting a `git clone` or `wget` or manual download of this repository, e.g. `git clone https://github.com/justincbagley/MAGNET.git`, then 
2.  granting the main script execute privileges by doing `chmod +x ./MAGNET` or `chmod +x ./MAGNET.sh`, then
3.  installing the dependencies (see below).

#### C. See also: PIrANHA Homebrew install

**MAGNET can also be installed by installing PIrANHA** as described in the [PIrANHA README](https://github.com/justincbagley/piranha) and [Wiki pages](https://github.com/justincbagley/piranha/wiki).

## SOFTWARE DEPENDENCIES

MAGNET v1.2.0 is a software package composed of `shell`, `R`, and Perl scripts and also calls several software programs that it relies on as dependencies. These dependencies are described in some detail in README files for different scripts in the package. However, here I provide a list of them. Dependencies that come standard on UNIX/Linux distributions, such as Perl 5+ or Python, are not listed.

<!--
- Perl (available at: https://www.perl.org/get.html).
- Nayoki Takebayashi's file conversion Perl scripts (available at: http://raven.iab.alaska.edu/~ntakebay/teaching/programming/perl-scripts/perl-scripts.html).
- Python (available at: https://www.python.org/downloads/).
-->
-   **bioscripts.convert v0.4 Python package**
    -   Description: 
        -   http://www.agapow.net/software/bioscripts.convert
    -   Installation: 
        -   Python Package Index: https://pypi.python.org/pypi/bioscripts.convert/0.4; also see README for `NEXUS2gphocs.sh`).
        -   `pip` install code: `pip install bioscripts.convert==0.4`

-   **RAxML - Randomized Axelerated Maximum Likelihood**
    -   Description:
        -   The Exelixis Lab RAxML page - https://cme.h-its.org/exelixis/web/software/raxml/
    -   Installation: 
        -   Source code: https://github.com/stamatak/standard-RAxML
        -   Conda install page: https://anaconda.org/bioconda/raxml
        -   Conda install code: `conda install -c bioconda raxml` (requires one of the Anaconda data science/Python distribution platform: https://www.anaconda.com/products/individual)

Users must install all software not included in MAGNET, and ensure that it is available via the command line on their local machine. On the user's local machine, Perl should be available by simply typing "`Perl`" or "`perl`" at the command line; Python should be available by typing "`python`" at the command line; and bioscripts.convert package should be available by typing "`convbioseq`" at the command line. Also, RAxML should be compiled using SSE3 install commands, so that RAxML can be called by simply typing "`raxmlHPC-SSE3`" on the command line. For detailed instructions for setting up RAxML this way, refer to the newest RAxML user manual (available at: http://sco.h-its.org/exelixis/resource/download/NewManual.pdf).

## INPUT FILE FORMAT

MAGNET assumes that you are starting from multilocus DNA sequence data in one of three formats. 

The **_first format_** that is supported is NEXUS format, with data in a single file having the extension '.nex'. 

The **_second format_** that is supported is that of a single datafile in G-Phocs (Gronau et al., 2011) format, with the extension '.gphocs'. 

For genomic data in aligned sequence format, such as aligned RAD tags (e.g. ddRAD-seq contigs) or other SNP data derived from genotyping-by-sequencing (GBS) methods, the user should assemble the data, call SNPs, and output SNP sequence data files in '.gphocs' or '.nex' format prior to running MAGNET. This can easily be done by running `pyRAD` or `ipyrad` (Eaton 2014) while calling for output in all formats (you'll get '.gphocs' and '.nex' files). 

However, this may not always be possible, and .gphocs format is not yet among the most popular file formats in phylogenomics/population genomics. Thus, I have added a `NEXUS2gphocs.sh` shell script utility within MAGNET (in the "shell" folder) that will convert a sequential NEXUS file into .gphocs format for you. An example NEXUS file "example.nex" is included in the distribution. Feel free to use the `NEXUS2gphocs.sh` utility script independently of MAGNET to convert from NEXUS to .gphocs format. However, when doing this, _make sure to follow the usage guidelines below_.

The **_third format_** that is supported in MAGNET is that of DNA sequence alignments for multiple loci contained in separate PHYLIP files for each locus. 

Users must specify the input fileType with the `-f` flag. Options are as follows: 

-   `1` for a single NEXUS-formatted input file (`<inputNEXUS>`, with extension '.nex'; also accepts a G-PhoCS-formatted input file, with extension '.gphocs'), or 
-   `2` for the multiple PHYLIP option. 

If `-f 1`, then the program will expect as standard input (stdin) the name of the `<inputNEXUS>`. However, if `-f 2`, then MAGNET expects to encounter multiple PHYLIP files to run on in the current working directory.

## PIPELINE

Apart from input file conversion steps, the MAGNET pipeline works by calling five different scripts, in series, each designed to conduct a task whose output is processed in the next step of the pipeline. First, the `gphocs2multiPhylip.sh` shell script is used to extract loci from the input file and place each locus in a PHYLIP-formatted file with extension '.phy'. Second, a shell script named `MultiRAxMLPrepper.sh` is used to place the .phy files into separate "run folders" and prepare them to be run in RAxML. Third, a script named `RAxMLRunner.sh` is called to run RAxML on the contents of each run folder. In a 'clean-up' step, MAGNET moves all '.phy' files files remaining in the working directory to a new folder, `phylip_files/`, which is created within the working directory.

After running the MAGNET pipeline, the shell script `getGeneTrees.sh` automates post-processing of the gene trees output by RAxML, including organizing all inferred gene trees into a single `gene_trees` folder in the working directory, and combining the individual 'best' gene trees resulting from each run into a single file named "besttrees.tre". Also, if bootstrap pseudoreplicates were performed and the bootstrap tree files are detected, then the `getBootTrees.sh` script conducts similar processing on the bootstrap trees for each loucus, which are collated, renamed, and given a list file containing the name of each file. Given the directory of bootstrap trees resulting from a MAGNET run (`bootstrap_trees`) can take up substantial disk space (>200 MB), users may wish to compress this directory to a zip file, for example using `$ zip -r bootstrap_trees.zip bootstrap_trees/` at the conclusion of a run.

A new feature of MAGNET (as of December 2018) is the `<resume>` option (`-R` and `--resume` flags; e.g. `-R 1` or `--resume 1`), which allow the user to resume a previous MAGNET run in the current working directory.

## USAGE

Additional input file and usage information is available in the usage or help texts. To get regular usage info for MAGNET, call for the help text by typing `./MAGNET -h` or `./MAGNET --help` (long option flag) at the terminal command line, and pressing enter. However, it is more useful (particularly when running for the first time, or after not running for awhile) to get _verbose usage information_ for MAGNET, including detailed descriptions of each option. You can do this by entering `./MAGNET -H` or `./MAGNET --Help` (long option flag) at the command line. 

The verbose usage text is as follows:

```txt

$ ./MAGNET -H

Usage: MAGNET [OPTION]...

 Options:
  -f, --filetype     fileType (def: 1; also: 2) starting file type; if 1, script expects as
                     stdin a single input NEXUS file in the current directory; if 2, then
                     script expects multiple input PHYLIP files in current directory
  -i, --input        inputNEXUS (def: NULL) input NEXUS file (mandatory for -f 1)
  -e, --exec         executable (def: raxml) name of RAxML executable available
                     from user's command line interface
  -b, --boot         numBootstraps (def: 100) RAxML bootstrap pseudoreplicates
  -r, --raxmlmodel   raxmlModel (def: GTRGAMMA; other: GTRGAMMAI, GTRCAT, GTRCATI)
  -s, --simplemodel  simpleModel (def: NULL; other: JC69, K80, HKY85) specifies 
                     simple DNA substitution model that will override any other model (even 
                     across partitions)
  -g, --gapthresh    gapThreshold (def: 0.001=essentially zero gaps allowed unless 
                     >1000 individuals; takes float proportion value) gap threshold value
  -m, --missing      indivMissingData (def: 1=allowed; 0=removed) missing data setting
  -o, --outgroup     outgroup (def: NULL) outgroup given as single taxon name (tip label) or
                     comma-separted list
  -h, --help         echo this help text and exit
  -H, --Help         echo verbose help text and exit
  -V, --version      echo version and exit
  -R, --resume       resume (def: 0, off; 1, on) option allowing user to resume a previous 
                     MAGNET run in the current working directory
  -d, --debug        debug (def: 0, off; 1, on) run function in Bash debug mode

 OVERVIEW
 The goal of MAGNET is to infer a maximum-likelihood (ML) gene tree in RAxML for each of 
 multiple loci, starting from one or multiple DNA sequence alignment input files. If supplied
 with a single G-PhoCS ('*.gphocs') or NEXUS ('*.nex') data file (using -f1 or -i <inputNEXUS> 
 -f1 options), then each locus is split into a separate PHYLIP alignment file, and RAxML 
 (Stamatakis 2014) is run to infer gene trees for each locus. If a NEXUS datafile is supplied, 
 it is converted into G-PhoCS format (Gronau et al. 2011) while splitting loci into separate 
 interleaved sequence blocks based on information provided in a sets block at the end of the 
 NEXUS file (e.g. defined using 'charset' commands), which is mandatory. However, if -f2, then 
 the program will run in current directory, assuming it contains multiple PHYLIP-formatted 
 alignment files. Under this scenario, MAGNET will skip directly to running the PHYLIP files 
 in RAxML using user-specified options. 
	Sequence names may not include hyphen characters, or there could be issues. For detailed 
 information on MAGNET and its various dependencies, see 'README.md' file in the distribution 
 folder; however, it is key that dependencies are available from the command line interface. 
 Among the most important options is <resume> (-r|--resume, off by default), which tells the
 program to resume a previous MAGNET run in current directory, including detecting incomplete 
 RAxML run folders, and running RAxML without overwriting results from the previous run(s).

 DETAILS
 The -f flag (also --filetype) specifies the starting fileType. If -f 1, then the mandatory 
	input is the name or path to the corresponding <inputNEXUS> starting file, which is 
	passed using the -i|--input flag. If -f 2, then mandatory input is the name or path to 
	the working directory (type '.' for current directory, or supply a relative or absolute 
	path).
 
 The -i flag (also --input) passess the name of the input NEXUS file, <inputNEXUS> parameter, 
    to the program.

 The -e flag (also --exec) sets the name of the RAxML executable that will be called. The 
    default executable name is 'raxml', but the user may wish to change this to something 
    specific to their install or parallelization needs (e.g. 'raxmlHPC-PTHREADS-SSE3'). The
    default setting should work on local machine or supercomputing cluster installs. However, 
    this should be tested beforehand by entering 'raxml' at the command prompt. On some 
    version fo Linux this yields the following error message:

	   'raxml: error while loading shared libraries: libmpi.so.12: cannot open shared object 
	    file: No such file or directory'.

	If this occurs, then Open MPI related libraries are installed in a non-standard location 
	and you will need to add this location to your LD_LIBRARY_PATH, e.g.:

	'export LD_LIBRARY_PATH=/usr/local/openmpi-1.8.1/lib:'
 
	See the following URL: for more insight into this problem: https://stackoverflow.com/
    questions/14769599/mpi-error-loading-shared-libraries. However, simply using a different 
    raxml executable that does not rely on these libararies will also immediately solve the 
    problem. In my experience, just setting MAGNET to call the 'raxmlHPC' executable immed-
    iately solves this issue on Mac and Linux (so also try simply running MAGNET with '-e 
    raxmlHPC' or '--exec raxmlHPC').
  
 The -b flag sets the number of boostrap pseudoreplicates for RAxML to perform while 
    estimating the gene tree for each locus. The default is 100; remove bootstrapping by 
    setting to 0.

 The -r flag sets the RAxML model for each locus. This uses the full default GTRGAMMA model,
	and at present it is not possible to vary the model across loci. If you want to use HKY
	or K80, you will need to use the -s flag (below).

 The -s flag sets a simple RAxML model for each locus/partition, which will override any
	model set using the -r flag above and apply to all partitions. In the current version of 
	RAxML, it is possible to specify the JC69, K80, and HKY85 models as overrides. By default,
	this option is turned off and the model set under the -r flag is used instead.

 The following two options are available **ONLY** if you are starting from a NEXUS input file:

	The -g flag supplies a 'gap threshold' to an R script, which deletes all column sites in 
	the DNA alignment with a proportion of gap characters '-' at or above the threshold value. 
	If no gap threshold is specified, all sites with gaps are removed by default. If end goal
	is to produce a file for G-PhoCS, you  will want to leave <gapThreshold> at the default. 
	However, if the next step in your pipeline involves converting from .gphocs to other data 
	formats, you will likely want to set <gapThreshold> = 1 (e.g. before converting to PHYLIP 
	format for RAxML). 

	The -m flag allows users to choose their level of tolerance for individuals with missing
	data. The default is <indivMissingData> = 1, allowing individuals with runs of 10 or more 
	missing nucleotide characters ('N') to be kept in the alignment. Alternatively, setting
	<indivMissingData> = 0 removes all such individuals from each locus; thus, while the input
	file would have had the same number of individuals across loci, the resulting file could
	have varying numbers of individuals for different loci.

 The -o flag sets the outgroup exactly the same way as that described in the RAxML v8 user's
	manual, as a single name or as a comma-separated list with no spaces between taxon names. 

changeLog.md
# MAGNET Change Log

## MAGNET v1.2.0 (current official major version release)

-   **December 23, 2020:** Updated MAGNET README (and saved PDF), and also prepared and released official major version 1.2.0 release, updated 'changeLog.md'.
-   **December 22-23, 2020:** Linted MAGNET scripts to pinpoint minor errors and weaknesses in the code, and went through and edited the main MAGNET shell scripts `MAGNET` and `MAGNET.sh`, as well as all `shell/` subfolder scripts.
-   **December 19-21, 2020:** Updated the standalone `MAGNET.sh` script and pipeline files (`shell` and `R` scripts) in this repository to version last released with the current head of PIrANHA (MAGNET v1.1.1), and then updated MAGNET to v1.2.0 here and in PIrANHA repo through various changes. Changes include a full rewrite of code for manually parsing the options, as an overhaul of the usage texts, and various minor edits to essentially all files (including some stylistic changes and some code improvements).

...

## MAGNET v0.1.9 (official minor version release - several changes after 0.1.6 release)

-   **February 2019:** Updated MAGNET script by adding a getBipartTrees function to the MAGNET pipeline, which organizes RAxML bipartitions trees for each locus (= best ML trees with bootstrap proportions along nodes the corresponding bootstrap searches search; resulting from ```-f a -x```options, which are included in all MAGNET calls to RAxML). Edited header and script banner to be unofficially prepped for future official release versioning 0.1.9.
-   **December 2018:** Added new MAGNET script updated to include --resume option, fix step info output to screen, and set raxml executable name one of two ways after detecting machine type (```raxml``` on Mac, ```raxmlHPC-SS3``` on Linux/supercomputer).
-   **November 25 2018:** Added new 'RAxMLRunChecker.sh' script v1.0, which counts the number of completed RAxML runs during the course of, or after, a MAGNET pipeline run, and also collates information on the dataset (e.g. number of patterns) and run (e.g. run time, optimum likelihood) for each locus/partition.
-   **November 20 2018:** Updated MAGNET with edited 'MAGNET.sh' (now v0.1.7) and 'NEXUS2gphocs.sh' (now v1.3) scripts containing an important bug fix and some new code checking for whether the NEXUS to fasta file conversion succeeded.


shell/NEXUS2gphocs.sh
#!/bin/sh

##########################################################################################
#                           NEXUS2gphocs v1.5.2, December 2020                           #
                                    VERSION="v1.5.2"                                     #
# Author: Justin C. Bagley                                                               #
# Date: Created by Justin Bagley on/before Aug 29 13:12:45 2016 -0700.                   #
# Last update: December 23, 2020                                                         #
# Copyright (c) 2016-2020 Justin C. Bagley. All rights reserved.                         #
# Please report bugs to <jbagley@jsu.edu>.                                               #
#                                                                                        #
# Description:                                                                           #
# SHELL SCRIPT THAT AUTOMATES SUBSAMPLING EACH OF ONE TO MULTIPLE PHYLIP ALIGNMENT       #
# FILES DOWN TO ONE (RANDOM) SEQUENCE PER SPECIES (FOR SPECIES TREE ANALYSIS)            #
#                                                                                        #
##########################################################################################

############ SCRIPT OPTIONS
## OPTION DEFAULTS ##
MY_GAP_THRESHOLD=0.001
MY_INDIV_MISSING_DATA=1

############ CREATE USAGE & HELP TEXTS
Usage="
Usage: $(basename "$0") [Help: -h help H Help] [Options: -g m] [stdin:] <inputNexus> 
 ## Help:
  -h   help text (also: -help -H -Help)

 ## Options:
  -g   gapThreshold (def: $MY_GAP_THRESHOLD=essentially zero gaps allowed unless >1000 
       individuals; takes float proportion value)
  -m   indivMissingData (def: $MY_INDIV_MISSING_DATA=allowed; 0=removed)
  -h   help text (also: --help) echo this help text and exit
  -V   version (also: --version) echo version and exit

 OVERVIEW
 Reads in a single NEXUS datafile and converts it to '.gphocs' format for G-PhoCS software
 (Gronau et al. 2011). Sequence names may not include hyphen characters, or there will be 
 issues. For best results, update to R v3.3.1 or higher.

 The -g flag supplies a 'gap threshold' to an R script, which deletes all column sites in 
 the DNA alignment with a proportion of gap characters '-' at or above the threshold value. 
 If no gap threshold is specified, all sites with gaps are removed by default. If end goal
 is to produce a file for G-PhoCS, you  will want to leave gapThreshold at the default. 
 However, if the next step in your pipeline involves converting from .gphocs to other data 
 formats, you will likely want to set gapThreshold=1 (e.g. before converting to phylip 
 format for RAxML). 

 The -m flag allows users to choose their level of tolerance for individuals with missing
 data. The default is indivMissingData=1, allowing individuals with runs of 10 or more 
 missing nucleotide characters ('N') to be kept in the alignment. Alternatively, setting
 indivMissingData=0 removes all such individuals from each locus; thus, while the input
 file would have had the same number of individuals across loci, the resulting file could
 have varying numbers of individuals for different loci.

 Dependencies: Perl; R; and Naoki Takebayashi Perl scripts 'fasta2phylip.pl' and 
 'selectSites.pl' in working directory or available from command line (in your path).

 CITATION
 Bagley, J.C. 2020. MAGNET v1.2.0. GitHub package, Available at: 
	<http://github.com/justincbagley/MAGNET>.
 or
 Bagley, J.C. 2020. MAGNET v1.2.0. GitHub package, Available at: 
	<https://doi.org/10.5281/zenodo.596774>.
"

if [[ "$1" == "-V" ]] || [[ "$1" == "--version" ]]; then
	echo "$(basename "$0") $VERSION";
	exit
fi

############ PARSE THE OPTIONS
while getopts 'h:H:g:m:' opt ; do
  case $opt in
## Help texts:
	h) echo "$Usage"
       exit ;;
	H) echo "$Usage"
       exit ;;

## Datafile options:
    g) MY_GAP_THRESHOLD=$OPTARG ;;
    m) MY_INDIV_MISSING_DATA=$OPTARG ;;

## Missing and illegal options:
    :) printf "Missing argument for -%s\n" "$OPTARG" >&2
       echo "$Usage" >&2
       exit 1 ;;
   \?) printf "Illegal option: -%s\n" "$OPTARG" >&2
       echo "$Usage" >&2
       exit 1 ;;
  esac
done

############ SKIP OVER THE PROCESSED OPTIONS
shift $((OPTIND-1)) 
# Check for mandatory positional parameters
if [ $# -lt 1 ]; then
echo "$Usage"
  exit 1
fi
MY_NEXUS="$1"


echo "INFO      | $(date) |----------------------------------------------------------------"
echo "INFO      | $(date) | NEXUS2gphocs, v1.5.2 December 2020  (part of PIrANHA v0.4a4)   "
echo "INFO      | $(date) | Copyright (c) 2016-2020 Justin C. Bagley. All rights reserved. "
echo "INFO      | $(date) |----------------------------------------------------------------"

######################################## START ###########################################

############ STEP #1: SETUP VARIABLES
###### Set filetypes as different variables:
echo "INFO      | $(date) | Examining current directory, setting variables... "
	export MY_WORKING_DIR="$(pwd)";
	export CR=$(printf '\r');
	calc () {
	   	bc -l <<< "$@" ;
	}


############ STEP #2: GET NEXUS FILE & DATA CHARACTERISTICS, CONVERT NEXUS TO FASTA FORMAT
# Extract charset info from sets block at end of NEXUS file: 
	MY_NEXUS_CHARSETS="$(egrep "charset|CHARSET" "$MY_NEXUS" | \
	awk -F"=" '{print $NF}' | sed 's/\;/\,/g' | \
	awk '{a[NR]=$0} END {for (i=1;i<NR;i++) print a[i];sub(/.$/,"",a[NR]);print a[NR]}' | \
	sed 's/\,/\,'$CR'/g' | sed 's/^\ //g')";

# Count number of loci present in the NEXUS file, based on number of charsets defined.
# Also get corrected count starting from 0 for numbering loci below...
	MY_NLOCI="$(echo "$MY_NEXUS_CHARSETS" | wc -l)";
	export MY_CORR_NLOCI="$(calc "$MY_NLOCI" - 1)";

# This is the base name of the original nexus file, so you have it. This WILL work regardless of whether the NEXUS filename extension is written in lowercase or in all caps, ".NEX".
	MY_NEXUS_BASENAME="$(echo "$MY_NEXUS" | sed 's/\.\///g; s/\.[A-Za-z]\{3\}$//g')";

# Convert data file from NEXUS to fasta format using bioscripts.convert v0.4 Python package:
# However, if alignment is too long (>100,000 bp), then need to convert to fasta using my 
# script and then wrap to 60 characters with fold function (as suggested at stackexchange
# post URL: https://unix.stackexchange.com/questions/25173/how-can-i-wrap-text-at-a-certain-column-size).
# If this conversion failes because the alignment is too long, then the code to follow 
# will have nothing to work with. So, I am here adding a conditional quit if the fasta
# file is not generated.

#---------ADD IF/THEN CONDITIONAL AND MY OWN NEXUS2fasta SCRIPT HERE!!!!----------#

	convbioseq fasta "$MY_NEXUS" > "$MY_NEXUS_BASENAME".fasta ;
	MY_FASTA="$(echo "$MY_NEXUS_BASENAME".fasta | sed 's/\.\///g; s/\.nex//g')";
	
	# The line above creates a file with the name basename.fasta, where basename is the base name of the original .nex file. For example, "hypostomus_str.nex" would be converted to "hypostomus_str.fasta".
	# Check to make sure the fasta was created; if so, echo info, if not, echo warning and quit:
	if [[ -s "$MY_NEXUS_BASENAME".fasta ]]; then
		echo "INFO      | $(date) |          Input NEXUS was successfully converted to fasta format. Moving forward... "
	else
		echo "WARNING   | $(date) |          NEXUS to fasta file conversion FAILED! Quitting... "
		exit 1 ;
	fi


############ STEP #3: PUT COMPONENTS OF ORIGINAL NEXUS FILE AND THE FASTA FILE TOGETHER TO
############ MAKE A G-PhoCS-FORMATTED DATA FILE
# Make top (first line) of the G-Phocs format file, which should have the number of loci on the first line:
echo "$MY_NLOCI" | sed 's/[\ ]*//g' > gphocs_top.txt ;

echo "$MY_GAP_THRESHOLD" > ./gap_threshold.txt ;
	count=0
	(
		for j in $MY_NEXUS_CHARSETS; do
			echo "$j"
			charRange="$(echo "$j" | sed 's/\,//g')";
			echo "$charRange";
			export setLower="$(echo "$j" | sed 's/\-.*$//g')";
			export setUpper="$(echo "$j" | sed 's/[0-9]*\-//g' | sed 's/\,//g; s/\ //g')";

			**/selectSites.pl -s $charRange $MY_FASTA > ./sites.fasta ;
			
			**/fasta2phylip.pl ./sites.fasta > ./sites.phy ;

			# Need to make sure there is a space between the tip taxon name (10 characters as output
			# by the fasta2phylip.pl Perl script) and the corresponding sequence, for all tips. Use
			# a perl search and replace for this:

			perl -p -i -e 's/^([A-Za-z0-9\-\_\ ]{10})/$1\ /g' ./sites.phy ;

				# If .phy file from NEXUS charset $j has gaps in alignment, then call 
				# rmGapSites.R R script to remove all column positions with gaps from
				# alignment and output new, gapless phylip file named "./sites_nogaps.phy". 
				# If charset $j does not have gaps, go to next line of loop. We do the 
				# above by first creating a temporary file containing all lines in
				# sites.phy with the gap character:
				grep -n "-" ./sites.phy > ./gaptest.tmp ;
				
				# Next, we test for nonzero testfile, indicating presence of gaps in $j, 
				# using UNIX test operator "-s" (returns true if file size is not zero). 
				# If fails, cat sites.phy into file with same name as nogaps file that
				# is output by rmGapSites.R and move forward:
				if [[ -s ./gaptest.tmp ]]; then
					echo "Removing column sites in locus${count} with gaps. "
					R CMD BATCH **/rmGapSites.R ;
				else
			   		echo ""
			   		cat ./sites.phy > ./sites_nogaps.phy ;
				fi
					
			export phylip_header="$(head -n1 ./sites_nogaps.phy)";
			locus_ntax="$(head -n1 ./sites_nogaps.phy | sed 's/[\ ]*[.0-9]*$//g')";
			locus_nchar="$(head -n1 ./sites_nogaps.phy | sed 's/[0-9]*\ //g')";
			
        		if [[ "$MY_INDIV_MISSING_DATA" = "0" ]]; then
					sed '1d' ./sites_nogaps.phy | egrep -v 'NNNNNNNNNN|nnnnnnnnnn' > ./cleanLocus.tmp ;
					cleanLocus_ntax="$(cat ./cleanLocus.tmp | wc -l)";
					echo locus"$((count++))" "$cleanLocus_ntax" "$locus_nchar" > ./locus_top.tmp ;
					cat ./locus_top.tmp ./cleanLocus.tmp >> ./gphocs_body.txt ;
				else
					echo locus"$((count++))" "$locus_ntax" "$locus_nchar" > ./locus_top.tmp ;
					cat ./locus_top.tmp ./sites_nogaps.phy >> ./gphocs_body.txt ;
				fi

			rm ./sites.fasta ./sites.phy ;
			if [[ "$(ls -1 ./*.tmp 2>/dev/null | wc -l | sed 's/\ //g')" != "0"  ]]; then 
				rm ./*.tmp ; 
			fi
			rm ./sites_nogaps.phy ;
		done
	)

	grep -v "^[0-9]*\ [0-9]*.*$" ./gphocs_body.txt > ./gphocs_body_fix.txt ;
	cat ./gphocs_top.txt ./gphocs_body_fix.txt > "$MY_NEXUS_BASENAME".gphocs ;

	############ STEP #4: CLEANUP: REMOVE UNNECESSARY FILES
	if [[ -s ./gphocs_top.txt ]]; then rm ./gphocs_top.txt ; fi ;
	if [[ -s ./gap_threshold.txt ]]; then rm ./gap_threshold.txt ; fi ;
	if [[ -s ./gphocs_body.txt ]]; then rm ./gphocs_body.txt ; fi ;

echo "INFO      | $(date) | Successfully created a '.gphocs' input file from the existing NEXUS file... "
echo "-------------------------------------------------------------------------------------"
#
#
#
######################################### END ############################################

exit 0
